### PR TITLE
Support for .attrs factory function (props => attrs)

### PR DESCRIPTION
**Pull request by oliverlaz**
A follow up PR of the discussion in #1171. The changes here are based on `v4.0.3`.

I made the smoke tests and test suite pass on my side so I'm opening this PR for an early review.
In the next few days, I will try this build in one of my apps (just to make sure that I didn't break something else).

ping @probablyup, @Fer0x

### CODE REVIEW

File: src/models/StyledComponent.js [hunk 1]
```diff
@@ -203,21 +186,27 @@ export default function createStyledComponent(target: Target, options: Object, r
...
 
   // fold the underlying StyledComponent attrs up (implicit extend)
   const finalAttrs =
     // $FlowFixMe
-    isTargetStyledComp && target.attrs ? { ...target.attrs, ...attrs } : attrs;
+    isTargetStyledComp && target.attrs
+      ? (context: Context): Object => ({
+          // $FlowFixMe <- target.attrs
+          ...processAttrs(target.attrs, context),
+          ...processAttrs(attrs, context),
```
Comment (Fer0x) [inline src/models/StyledComponent.js]: Maybe creating of this function here and in constructor can be extracted as `createAttrsResolver` and use typings.

**Comment by Fer0x** [comment]
@oliverlaz Awesome work, I'm glad you found time to finish it.

File: src/models/StyledComponent.js (hunk 1 above)
Comment (oliverlaz) [reply src/models/StyledComponent.js]: Sure, I will update this PR later.

File: src/utils/attrsResolver.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+/**
+ * Creates new attributes resolver function based on given attrs configuration.
+ *
+ * When the resolver function is invoked, it will return a plain object
+ * with keys set as attribute names and their respective evaluated values.
+ */
+export default function createAttrsResolver(...attrs: Array<?Attrs>): AttrsResolver {
+  const nonNullAttrs: Array<Attrs> = attrs.filter(Boolean);
```
Comment (probablyup) [inline src/utils/attrsResolver.js]: This is too opinionated. We should only be ignoring `undefined`.

**Comment by probablyup** [comment]
Thanks so much for this @oliverlaz! I ended up [implementing this a different way](#2200) that's a bit less code and helps support some more future use cases, but credited you in the changelog 🙇

**Comment by oliverlaz** [comment]
Hey @probablyup, thank you! Looking forward to see #2200 landed on master :)

**Pull request closed without merging**